draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+
+The server can then ask for a file-like object associated with the upload. The
+server can then use this object to perform validation, retrieve table rows, and
+settle them in their final home in the server (as is done currently, directly
+with the file contents).
+
+Once the work is completed, the server can delete the upload collection from the
+database.
+
+### API design
+
+This RFC adds an `uploads` section to the Multinet API, which contains the
+following endpoints:
+
+- `POST /api/uploads` - creates a new upload object by selecting a random,
+non-preexisting SHA string, then creating an Arango collection keyed by the
+string. This endpoint returns the ID string.
+- `POST /api/uploads/<upload-id>/chunk?sequence=<sequence-number>` - uploads one
+chunk of data to the named upload object. This endpoint returns the sequence
+number, and may raise various status codes to indicate problems.
+- `POST /api/uploads/<upload-id>/finalize` - finalizes the IDed upload by
+preventing further calls to `POST /api/uploads/<upload-id>/chunk` and performing
+a checksum on the settled data.
+- `DELETE /api/uploads/<upload-id>` - removes the database collection associated
+with the given ID.
+
+### File-like interface to upload collection
+
+A Python class named `UploadReader` which implements the file-like interface
+for reading data from an upload collection will be used by the serverside to
+interact with the data uploaded to the database using the `/api/uploads`
+endpoints. It should be constructed with a string ID referring to an upload, and
+it should implement the `seek()`, `read()`, and `readline()` methods as though
+the various documents in the collection constitute a regular file as opened by
+Python's `open()` function.
+
+### Usage
+
+Currently, file uploads are done by file type using the `uploaders` API section.
+There, a file of the specified type is sent to a specific `POST
+/api/uploaders/<file-type>` endpoint, which in turn handles the full file data
+including validation and ingestion to an Arango database and collection.
+
+Under this RFC, file upload becomes detached from file ingestion. Now, the
+correct sequence of `/api/uploads` calls must be made in order to settle the
+data in the database in an intermediate form.
+
+Next, the client can invoke one of the uploader paths, which will in turn
+instantiate an `UploadReader` object. This file-like object can be used to
+perform validation and bulk ingestion into a Multinet data table proper. The
+main advantage of this usage is that an `UploadReader` does not exhaust its data
+iterator after one pass; it can be rewound by using the `.seek()` method. This
+means that both validation and ingestion can be performed without a memory
+blowup caused by exhausting an HTTP request body iterator into a string in
+memory.
+
+Finally, the client (or the uploader methods) can invoke the `DELETE` endpoint
+to remove the intermediate upload data, leaving the server in a clean state.
```

Comment:
It seems a bit odd to me to put the client in charge of deleting the old upload data, after it's been validated and re-stored, as it's more of an implementation detail. That is, all the client cares about is that it's finished uploading its file, at which point it calls the `finalize` endpoint, and expects that after that request finishes, the data it cares about is validated and stored. So, I think deleting the old upload data should just be the final step of the `finalize` process.

This raises another question, which is that currently, we're depending on the fact that the processing/validation of uploads is relatively quick (i.e. I don't think we expect calls to the `finalize` endpoint to be taking too long). If we start processing larger data, this may not be the case, at which point something like background processing of a dataset may be necessary.

Again, I don't think we need to do this now, just something to keep on the radar.

Reply:
> It seems a bit odd to me to put the client in charge of deleting the old upload data, after it's been validated and re-stored, as it's more of an implementation detail. That is, all the client cares about is that it's finished uploading its file, at which point it calls the `finalize` endpoint, and expects that after that request finishes, the data it cares about is validated and stored. So, I think deleting the old upload data should just be the final step of the `finalize` process.
> 

The problem is, the `finalize` call can't necessarily delete the data because we still need the "uploader" to do its thing. It would be possible to tell the `finalize` call the name of the uploader we want to run (which could then run the uploader and on success, perform the deletion), but that seems like poor software design, since the uploaders themselves might need to be passed data or be called in arbitrary ways. We'll have to figure out how to do this properly. (Maybe the uploader itself can perform the deletion as a final step.)

> This raises another question, which is that currently, we're depending on the fact that the processing/validation of uploads is relatively quick (i.e. I don't think we expect calls to the `finalize` endpoint to be taking too long). If we start processing larger data, this may not be the case, at which point something like background processing of a dataset may be necessary.
> 
> Again, I don't think we need to do this now, just something to keep on the radar.

Good point. You can file an issue if you want to, though this problem will likely be obvious when it occurs, and at that time we can file issues, discuss what to do, etc.

Reply:
In my response I forgot to mention: I agree that it's odd to have the client issue the request for deletion. It ought to be done automatically the moment the intermediate data is no longer needed (specifically, without any extra input from the client).

Reply:
> The problem is, the `finalize` call can't necessarily delete the data because we still need the "uploader" to do its thing. It would be possible to tell the `finalize` call the name of the uploader we want to run (which could then run the uploader and on success, perform the deletion), but that seems like poor software design, since the uploaders themselves might need to be passed data or be called in arbitrary ways. We'll have to figure out how to do this properly. (Maybe the uploader itself can perform the deletion as a final step.)

I don't understand why the finalize call can't delete the data after uploads completes. To me, the finalize call has roughly the following steps:

1. Call the appropriate uploader by passing an instance of `UploadReader` (along with any necessary metadata) to the appropriate filetype uploader (the upload filetype could be either determined at finalization, or supplied by the client when initially creating the upload. I'm in favor of the latter.)

2. Once the previously called uploader completes, it has either validated and inserted that upload into the appropriate place, or has failed and thrown a validation error, which is just passed up the call stack and returned as the response to the client. If it's the former case, we've successfully inserted the uploaded data into its target location, without error, and can now delete the upload.

Reply:
I agree that having the client be responsible for deletion isn't ideal. I like @waxlamp's idea to have the uploader perform the deletion as the final step, although full disclosure I'm not 100% familiar with the uploaders yet, so I might be overlooking some issues with that approach.

Reply:
> > The problem is, the `finalize` call can't necessarily delete the data because we still need the "uploader" to do its thing. It would be possible to tell the `finalize` call the name of the uploader we want to run (which could then run the uploader and on success, perform the deletion), but that seems like poor software design, since the uploaders themselves might need to be passed data or be called in arbitrary ways. We'll have to figure out how to do this properly. (Maybe the uploader itself can perform the deletion as a final step.)
> 
> I don't understand why the finalize call can't delete the data after uploads completes. To me, the finalize call has roughly the following steps:
> 
> 1. Call the appropriate uploader by passing an instance of `UploadReader` (along with any necessary metadata) to the appropriate filetype uploader (the upload filetype could be either determined at finalization, or supplied by the client when initially creating the upload. I'm in favor of the latter.)

I also agree with using the latter.

> 2. Once the previously called uploader completes, it has either validated and inserted that upload into the appropriate place, or has failed and thrown a validation error, which is just passed up the call stack and returned as the response to the client. If it's the former case, we've successfully inserted the uploaded data into its target location, without error, and can now delete the upload.

Yeah, I like this idea.

Would the finalize call _require_ a filetype? That would mirror current practice where by necessity you have to select a file type in order to upload a file.

What do we do if the upload fails to pass validation? We have no way to alter the data on the server, so it seems like the only thing we can offer currently is to show the user the errors, and let them fix their data and try again. Once again, this mirrors current practice so I think it would be a fine place to start. In that case, we could also delete the upload on failure.

It would be a very advanced (and very cool) feature to allow in-place editing of uploaded data, but it's out of scope I think. We'd be entering shark-infested waters, since it's an open research/industry question of how to write a viable "data wrangler" tool.

Reply:
> Would the finalize call _require_ a filetype? That would mirror current practice where by necessity you have to select a file type in order to upload a file.

I thought the filetype would be specified on upload creation, is that correct?

> What do we do if the upload fails to pass validation? We have no way to alter the data on the server, so it seems like the only thing we can offer currently is to show the user the errors, and let them fix their data and try again. Once again, this mirrors current practice so I think it would be a fine place to start. In that case, we could also delete the upload on failure.

Yeah I think this is where we should start, as to not introduce too much complexity at once. However, I think it's important that we properly inform the user of the errors with an upload, so that it doesn't just fail for no reason.

> It would be a very advanced (and very cool) feature to allow in-place editing of uploaded data, but it's out of scope I think. We'd be entering shark-infested waters, since it's an open research/industry question of how to write a viable "data wrangler" tool.

I agree this would be a neat feature, and would require at least a few things:

* A data editor on the client to edit the uploaded data (and would differ by file-type)
* A paginated endpoint that returns the data from an upload
* An interface between the db and api layer to stream upload data (perhaps using the `UploadReader` class mentioned earlier.

However, we definitely shouldn't approach this to start. Just documenting my thoughts.